docs/internal/types.md

```diff
@@ -0,0 +1,359 @@
+This is in contrast to nominal typing, such as in Java, where an object implements an interface only if it adopts that interface.
+Objective-C supports both, with protocols defining nominal types but dynamic dispatch and explicit method checks providing structural typing.
+
+The 'algebraic' bit means that we can define types in terms of intersections or unions on other type expressions.
+For example, we can say `Integer | String` to define a set of constraints on a type that mean that it must either be an integer or a string (whatever they are: they may be concrete types, or they may be interfaces) or something like `Copyable & Hashable` to say any type that meets the constraints of the copyable and hashable interfaces.
+
+All types should be known at compile time and the compiler can assert that the objects and operations have the correct type at all times.
+Some types are inferred by the compiler, but there is no guessing: there's either one clear inferred type or it's an error.
+
+At a glance, Verona types are:
+ * Interfaces: abstract types that specify a set of fields and methods that an underlying concrete type must implement.
+   Interfaces may provide default implementations of methods that a class can adopt.
+ * Classes: concrete types.
+   A class may inherit from other classes but this does not make the class a subtype of the class from which it inherits.
+   Classes may explicitly adopt interfaces, in which case they adopt the default implementations from the interface, but conformance to an interface is entirely a property of the structural type.
+ * Capability rights: A Verona pointer is a *capability*: an unforgeable token of authority that permits access to an object. 
+   The set of rights that a capability holds are associated with a type, for example. `iso`, `mut`, `imm` and `readonly`.
+ * Unions: The concrete type of the object must fulfil all of the constraints of at least one out of two or more type expressions, for example the union type `(A | B)` must fulfil all of the requirements of either `A` or `B`.
+   Unions may refer to concrete types (for example, if `A` and `B` are classes then a variable of type `A | B` may be a reference to an instance of either class) or other type expressions, in which case the variable may refer to any concrete type that meets the requirements of either set of constraints.
+ * Intersections: The concrete type of the object must fulfil all of the constraints of two or more type expressions, for example the union type `(A & B)` must fulfil all of the requirements of both `A` and `B`, which may be interfaces or other type expressions.
+   No type satisfies the union type of two concrete types, so if `A` and `B` are both class types then `A` & `B` is a type that is not possible to satisfy.
+
+Capabilities offer language and runtime guarantees:
```

Comment:
Before diving into capabilities, it might be good to have a brief introduction to regions.

```suggestion
Verona manages memory by associating objects with 'regions'.
Each region has a single 'sentinel' object; all live objects in a region are reachable from its sentinel without crossing into another region.
To ensure data-race freedom, Verona limits programs to non-concurrent access to all objects of each region by carefully tracking the flow of references to the region's sentinel.

Capability types are the mechanism Verona uses to offer such language and runtime guarantees:
```